File: buddy_up_flutter/lib/data/models/analytics.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'analytics.freezed.dart';
part 'analytics.g.dart';

@freezed
abstract class AnalyticsUserInfo with _$AnalyticsUserInfo {
  const factory AnalyticsUserInfo({
    @Default('') String username,
    @JsonKey(name: 'display_name') @Default('') String displayName,
    @JsonKey(name: 'avatar_url') @Default('') String avatarUrl,
    @JsonKey(name: 'streak_days') @Default(0) int streakDays,
  }) = _AnalyticsUserInfo;

  factory AnalyticsUserInfo.fromJson(Map<String, dynamic> json) =>
      _$AnalyticsUserInfoFromJson(json);
}

@freezed
abstract class ActivityTypeBreakdown with _$ActivityTypeBreakdown {
  const factory ActivityTypeBreakdown({
    @Default('') String label,
    @Default(0) int count,
    @Default(0) double calories,
    @Default(0) double distance,
    @Default(0) double duration,
    @JsonKey(name: 'distance_km') @Default(0) double distanceKm,
  }) = _ActivityTypeBreakdown;

  factory ActivityTypeBreakdown.fromJson(Map<String, dynamic> json) =>
      _$ActivityTypeBreakdownFromJson(json);
}

@freezed
abstract class WorkoutRecent with _$WorkoutRecent {
  const factory WorkoutRecent({
    @JsonKey(name: 'performed_at') String? performedAt,
    @JsonKey(name: 'workout_type') @Default('') String workoutType,
    @Default('') String exercise,
    @JsonKey(name: 'duration_minutes') @Default(0) int durationMinutes,
    @JsonKey(name: 'calories_burned') double? caloriesBurned,
  }) = _WorkoutRecent;

  factory WorkoutRecent.fromJson(Map<String, dynamic> json) =>
      _$WorkoutRecentFromJson(json);
}

@freezed
abstract class WorkoutSummary with _$WorkoutSummary {
  const factory WorkoutSummary({
    @Default(0) int count,
    @JsonKey(name: 'total_calories_burned')
    @Default(0)
    double totalCaloriesBurned,
    @JsonKey(name: 'total_volume') @Default(0) double totalVolume,
    @JsonKey(name: 'by_type')
    @Default(<ActivityTypeBreakdown>[])
    List<ActivityTypeBreakdown> byType,
    @JsonKey(name: 'most_trained') String? mostTrained,
    @Default(<WorkoutRecent>[]) List<WorkoutRecent> recent,
  }) = _WorkoutSummary;

  factory WorkoutSummary.fromJson(Map<String, dynamic> json) =>
      _$WorkoutSummaryFromJson(json);
}

@freezed
abstract class ActivityRecent with _$ActivityRecent {
  const factory ActivityRecent({
    @Default('') String id,
    @JsonKey(name: 'activity_type') @Default('') String activityType,
    @JsonKey(name: 'started_at') String? startedAt,
    @JsonKey(name: 'duration_seconds') @Default(0) int durationSeconds,
    @JsonKey(name: 'distance_meters') @Default(0) double distanceMeters,
    @JsonKey(name: 'distance_km') @Default(0) double distanceKm,
    @JsonKey(name: 'avg_pace') double? avgPace,
    @JsonKey(name: 'calories_burned') double? caloriesBurned,
    @Default(<dynamic>[]) List<dynamic> route,
  }) = _ActivityRecent;

  factory ActivityRecent.fromJson(Map<String, dynamic> json) =>
      _$ActivityRecentFromJson(json);
}

@freezed
abstract class ActivitySummary with _$ActivitySummary {
  const factory ActivitySummary({
    @Default(0) int count,
    @JsonKey(name: 'total_distance_km') @Default(0) double totalDistanceKm,
    @JsonKey(name: 'total_duration_seconds')
    @Default(0)
    int totalDurationSeconds,
    @JsonKey(name: 'total_calories_burned')
    @Default(0)
    double totalCaloriesBurned,
    @JsonKey(name: 'total_steps') @Default(0) int totalSteps,
    @JsonKey(name: 'avg_pace') double? avgPace,
    @JsonKey(name: 'by_type')
    @Default(<ActivityTypeBreakdown>[])
    List<ActivityTypeBreakdown> byType,
    @Default(<ActivityRecent>[]) List<ActivityRecent> recent,
  }) = _ActivitySummary;

  factory ActivitySummary.fromJson(Map<String, dynamic> json) =>
      _$ActivitySummaryFromJson(json);
}

@freezed
abstract class MealRecent with _$MealRecent {
  const factory MealRecent({
    @Default('') String id,
    @JsonKey(name: 'meal_type') @Default('') String mealType,
    @JsonKey(name: 'food_name') @Default('') String foodName,
    @Default('') String description,
    double? calories,
    @JsonKey(name: 'protein_g') double? proteinG,
    @JsonKey(name: 'carbs_g') double? carbsG,
    @JsonKey(name: 'fat_g') double? fatG,
    @JsonKey(name: 'photo_url') @Default('') String photoUrl,
    @JsonKey(name: 'logged_at') String? loggedAt,
  }) = _MealRecent;

  factory MealRecent.fromJson(Map<String, dynamic> json) =>
      _$MealRecentFromJson(json);
}

@freezed
abstract class NutritionSummary with _$NutritionSummary {
  const factory NutritionSummary({
    @Default(0) int count,
    @JsonKey(name: 'total_calories') @Default(0) double totalCalories,
    @JsonKey(name: 'total_protein_g') @Default(0) double totalProteinG,
    @JsonKey(name: 'total_carbs_g') @Default(0) double totalCarbsG,
    @JsonKey(name: 'total_fat_g') @Default(0) double totalFatG,
    @JsonKey(name: 'by_type')
    @Default(<ActivityTypeBreakdown>[])
    List<ActivityTypeBreakdown> byType,
    @JsonKey(name: 'avg_daily_calories') double? avgDailyCalories,
    @Default(<MealRecent>[]) List<MealRecent> recent,
  }) = _NutritionSummary;

  factory NutritionSummary.fromJson(Map<String, dynamic> json) =>
      _$NutritionSummaryFromJson(json);
}

@freezed
abstract class BodySeriesPoint with _$BodySeriesPoint {
  const factory BodySeriesPoint({
    @Default('') String id,
    @JsonKey(name: 'weight_kg') @Default(0) double weightKg,
    @JsonKey(name: 'body_fat_pct') double? bodyFatPct,
    @JsonKey(name: 'measured_at') String? measuredAt,
    @JsonKey(name: 'photo_url') @Default('') String photoUrl,
    @JsonKey(name: 'scale_photo_url') @Default('') String scalePhotoUrl,
  }) = _BodySeriesPoint;

  factory BodySeriesPoint.fromJson(Map<String, dynamic> json) =>
      _$BodySeriesPointFromJson(json);
}

@freezed
abstract class BodySummary with _$BodySummary {
  const factory BodySummary({
    @Default(0) int count,
    @JsonKey(name: 'start_weight_kg') double? startWeightKg,
    @JsonKey(name: 'latest_weight_kg') double? latestWeightKg,
    @JsonKey(name: 'weight_change_kg') double? weightChangeKg,
    @JsonKey(name: 'latest_body_fat_pct') double? latestBodyFatPct,
    @Default(<BodySeriesPoint>[]) List<BodySeriesPoint> series,
  }) = _BodySummary;

  factory BodySummary.fromJson(Map<String, dynamic> json) =>
      _$BodySummaryFromJson(json);
}

@freezed
abstract class LivesSummary with _$LivesSummary {
  const factory LivesSummary({
    @JsonKey(name: 'joined_count') @Default(0) int joinedCount,
    @JsonKey(name: 'total_duration_seconds')
    @Default(0)
    int totalDurationSeconds,
    @JsonKey(name: 'by_type')
    @Default(<ActivityTypeBreakdown>[])
    List<ActivityTypeBreakdown> byType,
  }) = _LivesSummary;

  factory LivesSummary.fromJson(Map<String, dynamic> json) =>
      _$LivesSummaryFromJson(json);
}

@freezed
abstract class SpendingCategory with _$SpendingCategory {
  const factory SpendingCategory({
    @Default('') String category,
    @Default(0) int quantity,
    @Default(0) int count,
    @Default('') String label,
  }) = _SpendingCategory;

  factory SpendingCategory.fromJson(Map<String, dynamic> json) =>
      _$SpendingCategoryFromJson(json);
}

@freezed
abstract class SpendingSummary with _$SpendingSummary {
  const factory SpendingSummary({
    @JsonKey(name: 'gifts_sent')
    @Default(SpendingCategory())
    SpendingCategory giftsSent,
    @JsonKey(name: 'gifts_received')
    @Default(SpendingCategory())
    SpendingCategory giftsReceived,
    @JsonKey(name: 'tips_sent')
    @Default(SpendingCategory())
    SpendingCategory tipsSent,
    @JsonKey(name: 'tips_received')
    @Default(SpendingCategory())
    SpendingCategory tipsReceived,
    @JsonKey(name: 'live_fees')
    @Default(SpendingCategory())
    SpendingCategory liveFees,
    @JsonKey(name: 'gym_subscriptions')
    @Default(SpendingCategory())
    SpendingCategory gymSubscriptions,
    @JsonKey(name: 'session_fees')
    @Default(SpendingCategory())
    SpendingCategory sessionFees,
    @JsonKey(name: 'marketplace_spend')
    @Default(SpendingCategory())
    SpendingCategory marketplaceSpend,
    @JsonKey(name: 'total_transactions') @Default(0) int totalTransactions,
    @JsonKey(name: 'total_artifacts_spent') @Default(0) int totalArtifactsSpent,
    @Default(<SpendingCategory>[]) List<SpendingCategory> breakdown,
  }) = _SpendingSummary;

  factory SpendingSummary.fromJson(Map<String, dynamic> json) =>
      _$SpendingSummaryFromJson(json);
}

@freezed
abstract class ProgrammesSummary with _$ProgrammesSummary {
  const factory ProgrammesSummary({
    @JsonKey(name: 'programmes_purchased') @Default(0) int programmesPurchased,
    @JsonKey(name: 'meal_plans_purchased') @Default(0) int mealPlansPurchased,
    @JsonKey(name: 'active_enrolments') @Default(0) int activeEnrolments,
    @JsonKey(name: 'completed_enrolments') @Default(0) int completedEnrolments,
    @JsonKey(name: 'avg_progress_pct') double? avgProgressPct,
  }) = _ProgrammesSummary;

  factory ProgrammesSummary.fromJson(Map<String, dynamic> json) =>
      _$ProgrammesSummaryFromJson(json);
}

@freezed
abstract class AnalyticsSummaryData with _$AnalyticsSummaryData {
  const factory AnalyticsSummaryData({
    @Default('all') String period,
    @Default(AnalyticsUserInfo()) AnalyticsUserInfo user,
    @Default(WorkoutSummary()) WorkoutSummary workouts,
    @Default(ActivitySummary()) ActivitySummary activity,
    @Default(NutritionSummary()) NutritionSummary nutrition,
    @Default(BodySummary()) BodySummary body,
    @Default(LivesSummary()) LivesSummary lives,
    @Default(SpendingSummary()) SpendingSummary spending,
    @Default(ProgrammesSummary()) ProgrammesSummary programmes,
  }) = _AnalyticsSummaryData;

  factory AnalyticsSummaryData.fromJson(Map<String, dynamic> json) =>
      _$AnalyticsSummaryDataFromJson(json);
}

@freezed
abstract class AnalyticsReportResult with _$AnalyticsReportResult {
  const factory AnalyticsReportResult({
    @Default('') String id,
    @Default('all') String period,
    @Default(AnalyticsSummaryData()) AnalyticsSummaryData data,
    @JsonKey(name: 'image_url') @Default('') String imageUrl,
  }) = _AnalyticsReportResult;

  factory AnalyticsReportResult.fromJson(Map<String, dynamic> json) =>
      _$AnalyticsReportResultFromJson(json);
}

@freezed
abstract class ShareReportResult with _$ShareReportResult {
  const factory ShareReportResult({
    @JsonKey(name: 'report_id') @Default('') String reportId,
    @JsonKey(name: 'post_id') @Default('') String postId,
    @JsonKey(name: 'image_url') @Default('') String imageUrl,
  }) = _ShareReportResult;

  factory ShareReportResult.fromJson(Map<String, dynamic> json) =>
      _$ShareReportResultFromJson(json);
}

File: frontend/src/types/analytics.ts
export type AnalyticsPeriod = 'week' | 'month' | 'quarter' | 'year' | 'all';

export interface ActivityTypeBreakdown {
  activity_type?: string;
  label: string;
  count: number;
  distance?: number;
  distance_km?: number;
  duration?: number;
  calories?: number;
}

export interface WorkoutRecent {
  performed_at: string;
  workout_type: string;
  exercise: string;
  duration_minutes: number;
  calories_burned: number | null;
}

export interface WorkoutSummary {
  count: number;
  total_calories_burned: number;
  total_volume: number;
  by_type: ActivityTypeBreakdown[];
  most_trained: string | null;
  recent: WorkoutRecent[];
}

export interface ActivityRecent {
  id: string;
  activity_type: string;
  started_at: string | null;
  duration_seconds: number;
  distance_meters: number;
  distance_km: number;
  avg_pace: number | null;
  calories_burned: number | null;
  route: number[][];
}

export interface ActivitySummary {
  count: number;
  total_distance_km: number;
  total_duration_seconds: number;
  total_calories_burned: number;
  total_steps: number;
  avg_pace: number | null;
  by_type: ActivityTypeBreakdown[];
  recent: ActivityRecent[];
}

export interface MealRecent {
  id: string;
  meal_type: string;
  food_name: string;
  description: string;
  calories: number | null;
  protein_g: number | null;
  carbs_g: number | null;
  fat_g: number | null;
  photo_url: string;
  logged_at: string | null;
}

export interface NutritionSummary {
  count: number;
  total_calories: number;
  total_protein_g: number;
  total_carbs_g: number;
  total_fat_g: number;
  by_type: ActivityTypeBreakdown[];
  avg_daily_calories: number | null;
  recent: MealRecent[];
}

export interface BodySeriesPoint {
  id: string;
  weight_kg: number;
  body_fat_pct: number | null;
  measured_at: string | null;
  photo_url: string;
}

export interface BodySummary {
  count: number;
  start_weight_kg: number | null;
  latest_weight_kg: number | null;
  weight_change_kg: number | null;
  latest_body_fat_pct: number | null;
  series: BodySeriesPoint[];
}

export interface LivesSummary {
  joined_count: number;
  total_duration_seconds: number;
  by_type: ActivityTypeBreakdown[];
}

export interface SpendingCategory {
  category: string;
  quantity: number;
  count: number;
  label: string;
}

export interface SpendingSummary {
  gifts_sent: SpendingCategory;
  gifts_received: SpendingCategory;
  tips_sent: SpendingCategory;
  tips_received: SpendingCategory;
  live_fees: SpendingCategory;
  gym_subscriptions: SpendingCategory;
  session_fees: SpendingCategory;
  marketplace_spend: SpendingCategory;
  total_transactions: number;
  total_artifacts_spent: number;
  breakdown: SpendingCategory[];
}

export interface ProgrammesSummary {
  programmes_purchased: number;
  meal_plans_purchased: number;
  active_enrolments: number;
  completed_enrolments: number;
  avg_progress_pct: number | null;
}

export interface AnalyticsSummaryData {
  period: AnalyticsPeriod;
  user: {
    username: string;
    display_name: string;
    avatar_url: string;
    streak_days: number;
  };
  workouts: WorkoutSummary;
  activity: ActivitySummary;
  nutrition: NutritionSummary;
  body: BodySummary;
  lives: LivesSummary;
  spending: SpendingSummary;
  programmes: ProgrammesSummary;
}

export interface AnalyticsReportResult {
  id: string;
  period: AnalyticsPeriod;
  data: AnalyticsSummaryData;
  image_url: string;
}

export interface ShareReportPayload {
  period: AnalyticsPeriod;
  body?: string;
}

export interface ShareReportResult {
  report_id: string;
  post_id: string;
  image_url: string;
}

export interface ActivityRecordInput {
  activity_type: 'walk' | 'run' | 'hike' | 'cycle';
  started_at?: string;
  duration_seconds: number;
  distance_meters: number;
  avg_pace?: number | null;
  avg_speed_kmh?: number | null;
  calories_burned?: number | null;
  steps?: number | null;
  elevation_gain_m?: number | null;
  route?: number[][];
  notes?: string;
}

export interface WorkoutLogInput {
  workout_type: 'strength' | 'cardio' | 'hiit' | 'yoga' | 'mobility' | 'sport' | 'other';
  exercise?: string;
  sets?: number | null;
  reps?: number | null;
  weight_kg?: number | null;
  duration_minutes?: number;
  calories_burned?: number | null;
  distance_meters?: number | null;
  performed_at?: string;
  notes?: string;
}

export interface MealLogInput {
  meal_type: 'breakfast' | 'lunch' | 'dinner' | 'snack' | 'drink' | 'other';
  food_name?: string;
  description?: string;
  calories?: number | null;
  protein_g?: number | null;
  carbs_g?: number | null;
  fat_g?: number | null;
  photo_url?: string;
  logged_at?: string;
}

export interface BodyMetricInput {
  weight_kg: number;
  body_fat_pct?: number | null;
  photo_url?: string;
  notes?: string;
  measured_at?: string;
}
